ARROW-6599: [Rust] [DataFusion] Add aggregate traits and SUM implementation to physical query plan

andygrove opened the pull request:
This PR implements the traits for aggregate expressions and the SUM implementation.

paddyhoran commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
+impl Sum {
+    /// Create a new SUM aggregate function
+    pub fn new(expr: Arc<dyn PhysicalExpr>) -> Self {
+        Sum { expr }
```
nit: this can be `Self` instead of `Sum`

paddyhoran commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
+            DataType::UInt8 | DataType::UInt16 | DataType::UInt32 | DataType::UInt64 => {
+                Ok(DataType::UInt64)
+            }
+            other => Ok(other.clone()),
```
Can we be explicit here about what else is allowed?  i.e. floats should pass through but everything else should be an `Err` right?

paddyhoran commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
+macro_rules! sum_accumulate {
+    ($SELF:ident, $ARRAY:ident, $ROW_INDEX:expr, $ARRAY_TYPE:ident, $SCALAR_VARIANT:ident, $TY:ty) => {{
+        if let Some(array) = $ARRAY.as_any().downcast_ref::<$ARRAY_TYPE>() {
+            let value = array.value($ROW_INDEX);
```
You are not handling `NULL` slots here right?  Shouldn't you use [sum](https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/array_ops.rs#L75) from kernels module?  That way `NULL` handling is taken care of.  (SIMD, eventually, will be leveraged automatically as well).

andygrove replied:
This method is called to update an accumulator with the value from a single row within the batch, so SIMD can't help with this particular usage.

Basically I'm building a hash map of accumulators by grouping key, then iterating over the rows in the batch to look up the accumulator based on the grouping key in each row, then updating is.

There is a future optimization that can be made for aggregate queries without grouping keys e.g. `SELECT SUM(a) FROM b` where we could use the sum from kernels to sum entire batches.

paddyhoran replied:
Ok, I see. Yes, the example without grouping is what I had in mind.  Optimizations can be added later but are you handling `NULL` slots correctly here?

andygrove replied:
Good catch. I was not handling null slots correctly. I have pushed the fix and added a test with all null values.

paddyhoran commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
+                Some(ScalarValue::$SCALAR_VARIANT(n)) => {
+                    Some(ScalarValue::$SCALAR_VARIANT(n + value as $TY))
+                }
+                _ => Some(ScalarValue::$SCALAR_VARIANT(value as $TY)),
```
Shouldn't you match on `None` here, what else could it be?

paddyhoran commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
+        }
+        assert_eq!(accum.get_value()?, Some(ScalarValue::Int64(15)));
+
+        Ok(())
```
We should include a test that uses floats and contains nulls also.

andygrove commented:
Thanks for the review @paddyhoran ! I've addressed most of the comments. The reason for the concise function names for creating expressions is so that users can write concise code when building a query plan. For example:

```rust
HashAggregate:::new(
  col(0), // GROUP BY c0
  sum(col(1)), // SUM(c1)
  ..
);
```

Let me know what you think.

paddyhoran commented:
> The reason for the concise function names for creating expressions is so that users can write concise code when building a query plan.

Ok I see, in that context it makes sense to keep them short and concise.
